### DISCUSSION TITLE
Cypress after:spec is not being triggered whereas after:run is working fine. Cypress version: 12.5.1

### DISCUSSION THREAD

**Discussion by mgmpnandi** [cypress-io/cypress#27849] [Questions and Help]
Cypress version :12.5.1
after:run event getting triggered once the specs executions are finished but after:spec is not getting triggered at all.

Please help me urgently on this. Check my configuration below.

`import { defineConfig } from "cypress";
import { cloudPlugin } from "cypress-cloud/plugin";
let fs = require('fs');

export default defineConfig({
  defaultCommandTimeout: 40000,
  requestTimeout: 120000,
  numTestsKeptInMemory: 5,
  experimentalMemoryManagement: true,
  experimentalInteractiveRunEvents: true,
  retries: {
    runMode: 0,
    openMode: 0,
  },
  reporter: 'mochawesome',
  reporterOptions: {
    reportDir: 'cypress/results/my-test-output',
    overwrite: true,
    html: true,
    json: true,
    //timestamp: "MMddyyyy_HHMMss",
    reportFilename: "report-mochawesome"
  },
  e2e: {
    env: {},
    supportFile: "cypress/support/index.ts",
    testIsolation: false,
    setupNodeEvents(on, config) {
      
      //https://docs.cypress.io/api/plugins/after-run-api
      on('after:run', async (res) => {
        console.log("getting printed")
      })
      on('after:spec', async (spec, res) => {
        console.log("Not getting printed")
      })
      return cloudPlugin(on, config);
    }
  },
});
`

**Comment by portyu9** [comment]
Hi @mgmpnandi 

There are two things to check here.

First, if you are testing this with `cypress open`, `after:spec` does **not** fire immediately after each interactive spec execution. With `experimentalInteractiveRunEvents: true`, Cypress fires it when that spec's browser closes.

With `cypress run`, it should fire after each spec normally.

Second, your configuration has another likely conflict:

```js
on('after:spec', ...)
...
return cloudPlugin(on, config)
```

`cypress-cloud/plugin` also registers Cypress lifecycle events. Its own documentation warns that when multiple plugins register the same event, the later listener can replace the earlier listener. Since `cloudPlugin()` is initialized after your listener, your `after:spec` handler can be overwritten.

The workaround recommended by the plugin is to patch the `on` function, for example:

```js
import patchCypressOn from 'cypress-on-fix'
import { cloudPlugin } from 'cypress-cloud/plugin'

export default defineConfig({
  e2e: {
    async setupNodeEvents(cypressOn, config) {
      const on = patchCypressOn(cypressOn)

      on('after:spec', (spec, results) => {
        console.log('after:spec', spec.relative)
      })

      on('after:run', (results) => {
        console.log('after:run')
      })

      return cloudPlugin(on, config)
    },
  },
})
```

So if you are running headlessly, the `cypress-cloud` listener conflict is the first thing I would fix. If you are using `cypress open`, also remember that `after:spec` waits until the browser closes.

Please mark this answer as accepted if it helped, thank you.
